pkg/sql/mutations/mutations.go

```diff
@@ -122,14 +125,33 @@ func ForeignKeyMutator(rng *rand.Rand, stmts []tree.Statement) (additional []tre
 		// Check if a table has the needed column types.
 	LoopTable:
 		for refTable, refCols := range cols {
-			// Prevent self references.
-			// TODO(mjibson): Circular references are not
-			// prevented, but it would be nice to detect and
-			// prevent them.
+			// Prevent circular and self references because
+			// generating valid INSERTs could become impossible or
+			// difficult algorithmically.
 			if refTable == table.Table || len(refCols) < len(fkCols) {
 				continue
 			}
 
+			{
+				// To prevent circular references, find all transitive
+				// dependencies of refTable and make sure none of them
+				// are table.
+				stack := []tree.TableName{refTable}
+				for i := 0; i < len(stack); i++ {
```

Comment:
yes, that's why i did it this way. the `len` call will be evaluated anew each time. i don't know the semantics of how this would work with a range clause.

Reply:
makes sense